srcopsmetrics/metrics.py

```diff
@@ -0,0 +1,186 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with thoth-station/mi - Meta-information Indicators.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Metrics for MI."""
+
+import os
+from datetime import datetime
+from pathlib import Path
+
+import time
+import logging
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+from github import Github
+
+from srcopsmetrics.entities.issue import Issue
+from srcopsmetrics.entities.pull_request import PullRequest
+from srcopsmetrics.utils import check_directory
+
+from typing import Dict
+
+_LOGGER = logging.getLogger(__name__)
+
+_GITHUB_ACCESS_TOKEN = os.getenv("GITHUB_ACCESS_TOKEN")
+
+
+class Metrics:
+    """Metrics used in MI."""
+
+    def __init__(self, repository: str, visuaize: bool = False):
+        """Initialize with collected knowledge."""
+        gh_repo = Github(login_or_token=_GITHUB_ACCESS_TOKEN, timeout=50).get_repo(repository)
+
+        self.repo_name = repository
+        self.prs = PullRequest(gh_repo).load_previous_knowledge(is_local=True)
+        self.issues = Issue(gh_repo).load_previous_knowledge(is_local=True)
+        self.visualize = visuaize
+
+    def get_aggregated_pull_requests_with_metrics(self) -> pd.DataFrame:
+        """Aggregate analysed data, calculate metrics from it and return DataFrame."""
+        data = []
+        overall_ttfr = []
+        overall_ttm = []
+
+        for pr in self.prs.values():
+
+            if not pr["merged_at"]:
+                continue
+
+            created_at = int(pr["created_at"])
+
+            ttm = int(pr["merged_at"]) - created_at
+
+            reviewers = [pr["reviews"][r]["author"] for r in pr["reviews"]]
+            review_times = [int(pr["reviews"][r]["submitted_at"]) for r in pr["reviews"]]
+            ttfr = min(review_times) - created_at if review_times else None
+
+            reviews = [r for r in pr["reviews"].values()]
+            approvals = [r["submitted_at"] for r in reviews if r["state"] == "APPROVED"]
+            tta = min(approvals) - created_at if approvals else None
+
+            overall_ttfr.append(ttfr)
+            overall_ttm.append(ttm)
+
+            labels = [k for k in pr["labels"].keys()]
+
+            data.append([created_at, pr["created_by"], pr["size"], labels, reviewers, ttm, ttfr, tta])
+
+        aggregated = pd.DataFrame(data)
+        aggregated.columns = ["date", "author", "size", "labels", "reviewers", "ttm", "ttfr", "tta"]
+
+        factor = aggregated["ttm"]
+        normal = factor.between(factor.quantile(0.05), factor.quantile(0.95))
+        return aggregated[normal].sort_values(by=["date"]).reset_index(drop=True)
+
+    def save_graph_for_metrics(self, metrics_name: str, time_metrics_name: str):
+        """Calculate metrics score and plot graph."""
```

Comment:
this docstring also does not match method name, you evaluate metrics here as well and then you plot, there are different actions performed. If I want to improve the code or contribute and evaluate more metrics I'm not sure where to look for otherwise. wdyt?

Reply:
Yes that's right